include/boost/json/impl/monotonic_resource.ipp

```diff
@@ -12,59 +13,109 @@
+    block* next = nullptr;
+    unsigned char* base;
+    unsigned char* top;
 
     block(
-        std::size_t size_,
-        block* next_)
-        : size(size_)
-        , top(reinterpret_cast<
-            std::uintptr_t>(this+1))
-        , next(next_)
+        unsigned char* data,
+        std::size_t n,
+        block* head)
+        : size(n)
+        , next(head)
+        , base(data)
+        , top(data)
     {
     }
+};
 
-    void*
-    alloc(
-        std::size_t n,
-        std::size_t align) noexcept
+template<typename Block>
+void*
+monotonic_resource::
+allocate_in_block(
+    Block& b,
+    std::size_t size,
+    std::size_t align)
+{
+    const auto aligned = 
+        detail::align_up(b.top, align);
+    const auto next_top = aligned + size;
+    if(next_top <= b.base + b.size)
     {
-        // must be power of 2
-        BOOST_ASSERT(
-            (align & (align - 1)) == 0);
-        auto i = top;
-        if((i & (align - 1)) != 0)
-            i = (i | (align - 1)) + 1;
-        if(i + n > (reinterpret_cast<
-            std::uintptr_t>(this+1) + size))
-            return nullptr;
-        top = i + n;
-        return reinterpret_cast<void*>(i);
+        b.top = next_top;
+        return aligned;
     }
-};
+    return nullptr;
+}
 
 auto
 monotonic_resource::
-alloc_block(std::size_t size) ->
+allocate_new_block(std::size_t size) ->
     block&
+{  
+    /*-------------------------------------  
+             block storage layout:        
+           [ n ] = n allocated bytes
+      
+                header         data
+           [ sizeof(block) ] [ size ]
+      
+        data is guaranteed to be aligned
+          to alignof(std::max_align_t)
+    -------------------------------------*/
+    const auto bytes = size + sizeof(block);
+    const auto data = static_cast<unsigned char*>(
+        ::operator new(bytes));
+    head_ = ::new(static_cast<void*>(data))
+        block(data + sizeof(block), size, head_);
+    return *head_;
+}
+
+// returns the closest power of two
+// that is greater than requested
+std::size_t
+monotonic_resource::
+next_block_size(std::size_t requested)
 {
-    auto const n = (
-        size + sizeof(block) - 1) /
-        sizeof(block);
-    auto const bytes =
-        (n + 1) * sizeof(block);
-    auto& b = *::new(
-        ::operator new(bytes)) block(
-            n * sizeof(block), head_);
-    head_ = &b;
-    return b;
+    if(requested >= max_block_size_ >> 1)
+        return max_block_size_;
+    if(requested <= min_block_size_)
+        return min_block_size_;
+    std::size_t next = 1;
+    for(; requested; requested >>= 1, next <<= 1);
+    return next;
+}
+
+// skips rounding if requested
+// is already a power of two
+std::size_t
+monotonic_resource::
+closest_block_size(std::size_t requested)
+{
+    if(requested & (requested - 1))
+        return next_block_size(requested);
+    return requested;
+}
+
+std::size_t
+monotonic_resource::
+grow_block_size(std::size_t size)
```

Comment:
This is 112% better than overloads with `true_type` and `false_type` !!

Reply:
Epic